### PR TITLE
fix: Users can just create OperandRequest to install operator and CR …

### LINKED ISSUE #636: Creating Operator CRs by OperandRequest
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

```
apiVersion: operator.ibm.com/v1alpha1
kind: OperandRequest
metadata:
  name: db2-instance1
  namespace: my-service
spec:
  requests:
  - registry: common-service
    namespace: ibm-common-services
    operands:
    - name: ibm-db2-operator
      kind: db2
      instanceName: db2-instance1
      spec:
         replicas: 3
    - name: ibm-db2-operator
      kind: db2
      instanceName: db2-instance2
      spec:
         replicas: 3
```

This CR will install DB2 operator and create two DB2 instances. `db2-instance1` and `db2-instance2` into namespace `my-service`

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by horis233**
…and remove the dependencies to CSV

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #636

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by horis233** [comment]
@bitscuit @Daniel-Fan 

This PR is ready to review.

File: controllers/operandrequest/reconcile_operator.go [hunk 1]
```diff
@@ -213,10 +213,19 @@ func (r *Reconciler) deleteSubscription(ctx context.Context, operandName string,
...
+	originalsub := sub.DeepCopy()
 	if apierrors.IsNotFound(err) {
 		klog.V(3).Infof("There is no Subscription %s or %s in the namespace %s", operandName, op.PackageName, namespace)
 		return nil
+	} else if err != nil {
+		klog.Errorf("Failed to get Subscription %s or %s in the namespace %s", operandName, op.PackageName, namespace)
+		return err
+	}
+
+	if sub == nil || sub.Labels == nil {
```
Comment (Daniel-Fan) [inline controllers/operandrequest/reconcile_operator.go:225]: If the label is not existing, indicating it is not managed by ODLM.
If `sub` is `nil`, does it mean that the subscription is not exists and `apierrors.IsNotFound(err)` is `true`?

File: controllers/operandrequest/reconcile_operand.go [hunk 2]
```diff
@@ -200,69 +200,57 @@ func (r *Reconciler) reconcileCRwithConfig(ctx context.Context, service *operato
...
-		if operand.InstanceName == "" {
-			name = requestKey.Name
-		} else {
-			name = operand.InstanceName
-		}
+	var name string
+	if operand.InstanceName == "" {
+		name = requestKey.Name
+	} else {
+		name = operand.InstanceName
```
Comment (Daniel-Fan) [inline controllers/operandrequest/reconcile_operand.go:221]: Just wanna to confirm that the CR's name is `instanceName` or name of `OperandRequest`.
The `Spec.Requests[0].Operands[0].Name` is only for operator name.

Is it correct?

Comment (horis233) [reply controllers/operandrequest/reconcile_operand.go:221]: Yes and no.

`Spec.Requests[0].Operands[0].Name` is not for the operator name.  It is a common name to connect OperandRequest, OperandRegistry and OperandConfig. The operator name is defined in the OperandRegistry.

File: controllers/operandrequest/reconcile_operator.go (hunk 1 above)
Comment (horis233) [reply controllers/operandrequest/reconcile_operator.go:225]: Yes, I can modify it to `if sub.Labels == nil {`

File: controllers/operandrequest/reconcile_operand.go (hunk 2 above)
Comment (Daniel-Fan) [reply controllers/operandrequest/reconcile_operand.go:221]: yeah, get it. Others look good to me

**Review by Daniel-Fan — APPROVED** [review]

**Review by bitscuit — CHANGES_REQUESTED** [review]
Need to update design doc https://github.com/IBM/operand-deployment-lifecycle-manager/blob/master/docs/design/operand-deployment-lifecycle-manager.md. Not sure if there are any other places for doc to be updated as well.

**Comment by horis233** [comment]
> Need to update design doc https://github.com/IBM/operand-deployment-lifecycle-manager/blob/master/docs/design/operand-deployment-lifecycle-manager.md. Not sure if there are any other places for doc to be updated as well.

Yes, we also need a separate document for it

File: controllers/operandrequest/reconcile_operand.go [hunk 3]
```diff
@@ -200,69 +200,57 @@ func (r *Reconciler) reconcileCRwithConfig(ctx context.Context, service *operato
 }
 
 // reconcileCRwithRequest merge and create custom resource base on OperandRequest and CSV alm-examples
-func (r *Reconciler) reconcileCRwithRequest(ctx context.Context, requestInstance *operatorv1alpha1.OperandRequest, operand operatorv1alpha1.Operand, requestKey types.NamespacedName, csv *olmv1alpha1.ClusterServiceVersion) error {
-	almExamples := csv.GetAnnotations()["alm-examples"]
```
Comment (bitscuit) [inline controllers/operandrequest/reconcile_operand.go:204]: So are we no longer using alm-examples to create CR? Does this mean we cannot use the default CR provided by operators' CSVs?

**Review by bitscuit — CHANGES_REQUESTED** [review]

**Comment by bitscuit** [comment]
> 
> 
> > Need to update design doc https://github.com/IBM/operand-deployment-lifecycle-manager/blob/master/docs/design/operand-deployment-lifecycle-manager.md. Not sure if there are any other places for doc to be updated as well.
> 
> Yes, we also need a separate document for it

Separate document where and why? We just need to update the current doc, right?

**Comment by horis233** [comment]
I expect a new feature doc because it implements a new install path instead of an enhancement.

File: controllers/operandrequest/reconcile_operand.go (hunk 3 above)
Comment (horis233) [reply controllers/operandrequest/reconcile_operand.go:204]: This is only for creating CR by OperandRequest. We still support creating CRs by OperandConfig

File: docs/design/create-cr-by-operandrequest.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+```yaml
+apiVersion: operator.ibm.com/v1alpha1
+kind: OperandRequest
+metadata:
+  name: db2-instance1
+  namespace: my-service
+spec:
+  requests:
+  - registry: common-service
+    namespace: ibm-common-services
```
Comment (bitscuit) [inline docs/design/create-cr-by-operandrequest.md:22]: should this be `registryNamespace`?

File: docs/design/create-cr-by-operandrequest.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        replicas: 3
+    - name: ibm-db2-operator
+      kind: db2
+      apiVersion: operator.ibm.com/v1alpha1
+      instanceName: db2-instance2
+      spec:
+        replicas: 3
+```
+
+The above `OperandRequest` will create two `db2` custom resources in the `my-service` namespace and the `db2` operator will be created in the namespace specified in the OperandRegistry `ibm-common-services/common-services`.
```
Comment (bitscuit) [inline docs/design/create-cr-by-operandrequest.md:38]: unclear where `db2` operator will be created. Is it deployed in `ibm-common-services` or `common-services`?

**Review by bitscuit — CHANGES_REQUESTED** [review]

File: docs/design/create-cr-by-operandrequest.md (hunk 4 above)
Comment (horis233) [reply docs/design/create-cr-by-operandrequest.md:22]: Thanks, nice catch

File: docs/design/create-cr-by-operandrequest.md (hunk 5 above)
Comment (horis233) [reply docs/design/create-cr-by-operandrequest.md:38]: Let me make it clear. I mean it the namespace info is in the OperandRegistry `common-service` in the `ibm-common-services`, Let me update it to make it clear.

**Review by bitscuit — APPROVED** [review]
/lgtm

**Pull request merged**